src/Sylius/Component/Core/OrderProcessing/StateResolver.php

```diff
@@ -61,35 +79,48 @@ public function resolvePaymentState(OrderInterface $order)
      */
     public function resolveShippingState(OrderInterface $order)
     {
-        $order->setShippingState($this->getShippingState($order));
-    }
+        if (OrderShippingStates::SHIPPED === $order->getShippingState()) {
+            return;
+        }
+        /** @var Collection $shipments */
+        $shipments = $order->getShipments();
+        $shipmentCount = $shipments->count();
+        $shipmentInShippedStateAmount = $shipments->filter(
+            function (ShipmentInterface $shipment) {
+                return ShipmentInterface::STATE_SHIPPED === $shipment->getState();
+            })
+            ->count()
+        ;
+        $shipmentInReadyStateAmount = $shipments->filter(
+            function (ShipmentInterface $shipment) {
+                return ShipmentInterface::STATE_READY === $shipment->getState();
+            })
+            ->count()
+        ;
+        $shipmentInCancelledStateAmount = $shipments->filter(
+            function (ShipmentInterface $shipment) {
+                return ShipmentInterface::STATE_CANCELLED === $shipment->getState();
+            })
+            ->count()
+        ;
 
-    /**
-     * @param OrderInterface $order
-     *
-     * @return string
-     */
-    protected function getShippingState(OrderInterface $order)
-    {
-        $states = [];
+        /** @var StateMachine $stateMachine */
+        $stateMachine = $this->stateMachineFactory->get($order, OrderShippingTransitions::GRAPH);
 
-        foreach ($order->getShipments() as $shipment) {
-            $states[] = $shipment->getState();
+        if ($shipmentCount === $shipmentInShippedStateAmount && OrderShippingStates::SHIPPED !== $order->getShippingState()) {
```

Comment:
To lower the cognitive load we can extract this condition as `$this->allShipmentsShippedButOrderStateNotUpdated()` or something similar.